packages/host/app/components/ai-assistant/formatted-message.gts

```diff
@@ -284,72 +274,66 @@ function parseHtmlContent(htmlString: string): HtmlTagGroup[] {
-      content: '',
-    };
-  }
-  // Extract language using regex - finds the value of data-codeblock attribute
-  const languageMatch = preElementString.match(/data-code-language="([^"]+)"/);
-  const language = languageMatch ? languageMatch[1] : null;
-
-  // Extract content using regex - finds everything between the opening and closing pre tags
-  const contentMatch = preElementString.match(/<pre[^>]*>([\s\S]*?)<\/pre>/);
-  let content = contentMatch ? contentMatch[1] : null;
-
-  // When the model responds with a code patch block, the source-code-editing.json skill
-  // will dictate that the first comment in the search/replace file block
-  // is a file url comment, like this: // File url: http://localhost:4201/user/realm/card.gts
-  // We need this url so that we know which file the code patch applies to.
-  // We extract this url and remove it from the content.
-  let fileUrl = null;
-
-  let trimmedContent = content?.trim();
-  if (trimmedContent?.trim().startsWith('// File url: ')) {
-    let firstLine = trimmedContent.split('\n')[0];
-    let fileUrlRegex = /File url: (.*)/;
-    let fileUrlMatch = firstLine.match(fileUrlRegex);
-    if (fileUrlMatch) {
-      fileUrl = fileUrlMatch[1];
-    }
-    // Remove the first line from the content
-    content = trimmedContent.split('\n').slice(1).join('\n');
-  }
-
-  return {
-    language: language ?? '',
-    content: content ?? '',
-    fileUrl: fileUrl ?? null,
+interface CodeDiffResourceArgs {
+  named: {
+    fileUrl?: string | null;
+    searchReplaceBlock?: string | null;
   };
 }
 
-function findLastTextNodeWithContent(parentNode: Node): Text | null {
-  // iterate childNodes in reverse to find the last text node with non-whitespace text
-  for (let i = parentNode.childNodes.length - 1; i >= 0; i--) {
-    let child = parentNode.childNodes[i];
-    if (child.textContent && child.textContent.trim() !== '') {
-      if (child instanceof Text) {
-        return child;
-      }
-      return findLastTextNodeWithContent(child);
-    }
+export class CodeDiffResource extends Resource<CodeDiffResourceArgs> {
+  @tracked fileUrl: string | undefined | null;
+  @tracked originalCode: string | undefined | null;
+  @tracked modifiedCode: string | undefined | null;
+  @tracked searchReplaceBlock: string | undefined | null;
+
+  @service private declare cardService: CardService;
+  @service private declare commandService: CommandService;
+
+  modify(_positional: never[], named: CodeDiffResourceArgs['named']) {
+    let { fileUrl, searchReplaceBlock } = named;
+    this.fileUrl = fileUrl;
+    this.searchReplaceBlock = searchReplaceBlock;
+
+    this.load.perform();
+  }
+
+  get isDataLoaded() {
+    return !!this.originalCode || !!this.modifiedCode;
   }
-  return null;
+
+  load = restartableTask(async () => {
+    let { fileUrl, searchReplaceBlock } = this;
+    if (!fileUrl || !searchReplaceBlock) {
+      return;
+    }
+    let result = await this.cardService.getSource(new URL(fileUrl));
```

Comment:
Here, and on line 317, it is entirely possible an error can happen, either due to failed fetch, or failed patch operation. I think it would be good to handle these and make it so that the assistant offers the user to retry. I added a ticket to add that.